src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -6730,6 +6750,126 @@ static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpress
 
     if ( ! relationFunction || ( geometry.*relationFunction )( feat2.geometry() ) ) // Calls the method provided as template argument for the function (e.g. QgsGeometry::intersects)
     {
+
+      if ( isIntersectsFunc && ( minOverlap != -1 || minInscribedCircleRadius != -1 ) )
+      {
+        const QgsGeometry intersection { geometry.intersection( feat2.geometry() ) };
+
+        switch ( intersection.type() )
+        {
+          case QgsWkbTypes::GeometryType::PolygonGeometry:
+          {
+            // Check min overlap for intersection (if set)
+            if ( intersection.isMultipart() )
+            {
+              bool testResult { false };
+              for ( auto it = intersection.const_parts_begin(); ! testResult && it != intersection.const_parts_end(); ++it )
```

Comment:
it's safe to iterate over parts for a singlepart geometry, so you could remove the separate branches for singlepart/multipart here and always use the iterator.